### ISSUE TITLE
error: cannot find derive macro `BitfieldStruct` in this scope

### ISSUE THREAD

**Issue by ctaggart** [immunant/c2rust#183]
I tried to manually fix it for about a half hour and wasn't able to, so I don't think it is trivial. (from https://github.com/ctaggart/openconnect-rust/issues/5)

``` rust
Camerons-MacBook-Pro:openconnect-rust cameron$ cargo build
   Compiling openconnect-rust v0.0.0 (/Users/cameron/github/openconnect-rust)
error: cannot find derive macro `BitfieldStruct` in this scope
    --> src/tun.rs:1198:29
     |
1198 |     #[derive ( Copy, Clone, BitfieldStruct )]
     |                             ^^^^^^^^^^^^^^

error: cannot find derive macro `BitfieldStruct` in this scope
    --> src/gpst.rs:1086:29
     |
1086 |     #[derive ( Copy, Clone, BitfieldStruct )]
     |                             ^^^^^^^^^^^^^^
```

It is complaining about:

https://github.com/ctaggart/openconnect-rust/blob/28ff0a3ca51386d841638c7fff9fe6a862e7465d/src/tun.rs#L1198-L1215

``` rust
    #[derive ( Copy, Clone, BitfieldStruct )]
    #[repr(C)]
    #[src_loc = "81:1"]
    pub struct ip {
        #[bitfield(name = "ip_hl", ty = "u_int", bits = "0..=3")]
        #[bitfield(name = "ip_v", ty = "u_int", bits = "4..=7")]
        pub ip_hl_ip_v: [u8; 1],
        pub ip_tos: u_char,
        pub ip_len: u_short,
        pub ip_id: u_short,
        pub ip_off: u_short,
        pub ip_ttl: u_char,
        pub ip_p: u_char,
        pub ip_sum: u_short,
        pub ip_src: in_addr,
        pub ip_dst: in_addr,
        /* source and dest address */
    }
```

**Comment by TheDan64** [comment]
I think your `use c2rust_bitfields::BitfieldStruct;` import needs to be moved into the `mod` block containing the struct. Rust modules don't inherit imports from parents.

**Comment by ctaggart** [comment]
Yes, I tried that yesterday. I tried that again and noticed it is actually a different error:

```
error: cannot determine resolution for the derive macro `BitfieldStruct`
    --> src/tun.rs:1199:29
     |
1199 |     #[derive ( Copy, Clone, BitfieldStruct )]
     |                             ^^^^^^^^^^^^^^
     |
     = note: import resolution is stuck, try simplifying macro imports

error: cannot find derive macro `BitfieldStruct` in this scope
    --> src/gpst.rs:1086:29
     |
1086 |     #[derive ( Copy, Clone, BitfieldStruct )]
     |                  
```

**Comment by TheDan64** [comment]
I did manage to figure out a way to get it work, using the old macro import approach:

In your `lib.rs`, after your feature attrs but before your mod decls, add:
```rust
#[macro_use]
extern crate c2rust_bitfields;
```

This will make the derive macro a globally visible type, so you can use it in the rest of the crate without any use imports.

That said, this shouldn't be a necessary - so I'm inclined to think that this is a rust macro resolution bug in the 2018e macro imports.

**Comment by rinon** [comment]
This was my mistake, I removed the `extern crate` for all things we use. We'll need to put that back in for proc-macro crates, unfortunately.

**Comment by rinon** [comment]
The import should work, but you'll need to import it again in the `ip_h` module in the file you linked to. This should be automatic, we need to fix the translator to do that.

**Comment by TheDan64** [comment]
@ctaggart Any luck with my workaround?

**Comment by rinon** [comment]
Since rustc is deprecating custom attributes and WONTFIX the issue above, we should revert back to using `extern crate`.

**Comment by TheDan64** [comment]
Do we need to revert back though? If custom attrs are going away, then this bug shouldn't appear? By that, I mean that the issue is due to the attr, which is going away/to be replaced by another approach

**Comment by TheDan64** [comment]
I guess we should do that if we want a more immediate fix, though

**Comment by rinon** [comment]
It's going to take a lot longer to work around the attributes thing. For now we'll keep using the custom attribute and just need to fix the imports.
